Feat/field permission: 접근할 수 있는 필드를 권한에 따라 조정

Ho-s opened the pull request:
## #️⃣ 연관된 이슈

> ex) #이슈번호, #이슈번호

Resolves #13 

### 📝 작업 내용

> 이번 PR에서 작업한 내용을 간략히 설명해주세요

- Graphql에서는 사용자가 원하는 대로 entity에 접근할 수 있기 때문에, 권한을 나눠줬어요.
- 기존의 prisma-nestjs-graphql 라이브러리는 커스텀 데코레이터를 넣을 수 없어서 삭제했습니다.

### 스크린샷 (선택)

## 💬 리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
>
> ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?

- 천천히 확인해주세요~


## 📚 참고할만한 자료(선택)

Ho-s committed:
feat: graphql context에서 user 정보 받도록

Ho-s committed:
feat: graphql 에서 유저가 선택한 필드 읽을 수 있는 권한을 검증

Ho-s committed:
chore: prisma-nestjs-graphql 의존성 삭제 -> 직접 작성하는 방식으로 변경

- custom decorator를 사용하기 위함

Ho-s committed:
chore: change to enum

Ho-s committed:
feat: user, post 의 graphql용 response dto를 설정

labyrinth30 commented on src/common/config/graphql.context.ts:
```diff
@@ -0,0 +1,25 @@
+    reply,
+    /**
+     * @todo
+     * Guard에서 주입받도록
```
todo 추가 좋습니다!

labyrinth30 commented on src/common/field-access/field-access.decorator.ts:
```diff
@@ -0,0 +1,11 @@
+    if (!roles.length) roles = ['USER'];
+    Reflect.defineMetadata(FIELD_ROLE, roles, target, propertyKey);
+  };
+};
```
👍

labyrinth30 commented on src/common/field-access/field-access.interceptor.ts:
```diff
@@ -0,0 +1,116 @@
+    return requiredRole?.some((role) => role === userRole) ?? true;
+  }
+
+  private extractSelecFields<T>(fieldNodes: FieldNode[]): (keyof T)[] {
```
여기 메서드 이름 오타 같은데 의도하신 건가요?

Ho-s replied:
엇!! 그냥 오타입니다..!! ㅋㅋㅋ 수정하겠습니다!

labyrinth30 commented on src/prisma/schema.prisma:
```diff
@@ -3,11 +3,6 @@ generator client {
   provider      = "prisma-client-js"
 }
 
-generator nestgraphql {
```
커스텀 데코레이터 사용 위해 의존성 제거한 부분 찾아보니 이해가 됩니다! 그래서 이후 커밋에서 dto들을 따로 작성하신 것이군요

Ho-s replied:
맞습니다! 커스텀 데코레이터를 넣을 수 없어서 직접 구현하였는데,
추후에 자동화하는 로직도 구성하면 재밌을 것 같습니다~!

labyrinth30 reviewed (approved):
상세한 커밋 감사합니다!
저도 프로젝트하면서 꼭 todo들 구현 진행하도록 하겠습니다:)

Ho-s committed:
fix: fix typo

Ho-s committed:
Update README.md

hye-on reviewed (approved):
열일 하시는군요
잘봤습니당 👍 고생했어!!

labyrinth30 committed:
Feat/guard (#9)

* feat: passport 사용한 guard 작업을 위한 라이브러리 설치

- @nestjs/jwt, @nestjs/passport, passport, passport-jwt, passport-local 설치
- @types/passport-jwt, @types/passport-local 설치

* feat: auth module 생성

* feat: bcrypt, class-validator 설치

* feat: auth module에 jwt module 추가 및 user module과 연결

- .example.env에 jwt 관련 환경변수 추가

* feat: LocalAuthGuard 구현

* feat: 로그인 후 반환되는 ObjectType JwtWithUser 구현

* feat: current-user.decorator.ts 구현

- gql 속 현재 user 정보를 가져오는 커스텀 데코레이터 구현

* feat: LocalAuthGuard에 사용되는 authenticate 메서드 구현

* chore: @nestjs/passport 버전 변경

- passport-jwt와의 호환성 문제 해결

* feat: JwtAuthGuard 구현

* feat: AuthModule에 AuthGuard들 추가

* feat: RS256에서 HS256으로 변경

* chore: AuthModule에 Strategy들 추가

* feat: signIn, signUp 구현

* chore: 400에러를 401에러로 구체화

* chore: @nestjs/passport 버전 복구

- ^10.0.3으로 다운그레이드 -> ^11.0.5로 복구

* chore: JwtStrategy 타입 이슈 해결

Ho-s committed:
chore: resolve conflict
